v2.2/transaction-retry-errors.md

```diff
@@ -0,0 +1,152 @@
+committed for some reason, it needs to make sure that everything that
+happened as a result of that transaction is undone.
+
+This includes reverting decisions made client-side using the data
+**returned by the database while the txn was open**, such as:
+
+- whether to advance the program counter in the driving application,
+- whether to send an email to a user,
+- whether to send a package through the postal service.
+
+In a traditional conversational SQL model, there’s no way for the
+database to guarantee that these things will be rolled back without
+informing the user that they need to do it themselves. In most cases
+this means requiring the application to not take irreversible actions
+until the transaction has successfully committed; but there is no way
+to enforce this in the SQL protocol.
+
+Some other SQL engines will avoid this problem by **preventing
+interactive transactions**, i.e. disallow a client from observing the
+result of an intermediate statement before it issues another statement
+while the txn is open. When a SQL engine only supports non-interactive
+transactions, it can freely decide to delay reporting the results
+until a txn is successfully committed, and handle txn retries
+internally itself. CockroachDB does not do this because it supports
+the PostgreSQL dialect which mandates support for interactive
+transactions.
+
+# Transaction scheduling
+
+What is causing hurdle (2) above?
+
+Shouldn't the database have a sophisticated transaction scheduler?
+
+If it had, it would intelligently provide concurrency, make
+transactions run as soon as theoretically possible, but prevent wrong
+orderings not allowed by serializability.
+
+The difficulty with this is twofold:
+
+1. To decide a good ordering requires a precise
+   analysis of the possible conflicts between transactions, also
+   called "dependency analysis". There are two main obstacles to being
+   able to do that in the first place:
+   - SQL is pretty expressive:
+     ```
+       SELECT x FROM a WHERE y = (SELECT q FROM b LIMIT 1)
+     ```
+     To determine the set of dependencies may require running a (sub-)query.
+   - For transactions containing more than 1 statement (interactive),
+     the set of dependencies is not computable, because the
+     transaction must start (and let some statements execute) before
+     subsequent statements start and their dependencies are known.
+2. Even if the dependencies were exactly known and it was
+   theoretically possible to compute a good ordering, this is a
+   computationally hard problem to solve (it's NP-complete) and thus
+   may not be practical to 2resolve while keeping latencies low.
+
+Because of this, most SQL engines (including CockroachDB, but also
+PostgreSQL) have a trade-off: they will **attempt** to compute
+dependencies in common cases when it is both possible and fast to do
+so, and make a **best effort** at organizing transaction ordering in a way
```

Comment:
I'm not sure what this paragraph is referring to exactly. What is our "best effort" at ordering transactions? It's not like we try different possible orderings.